src/bh_db_worker.erl

```diff
@@ -38,46 +48,75 @@ squery(Pool, Sql) ->
+                    receive
+                        {Conn, Ref, Res} ->
+                            Res
+                    after
+                        500 ->
+                            throw(?RESPONSE_503)
+                    end;
+                Error ->
+                    dispcount:checkin(Pool, Reference, Conn),
+                    lager:warning("failed to parse query ~p : ~p", [Stmt, Error]),
+                    throw({500, [], <<"Internal server error">>})
+            end;
         {error, busy} ->
             throw(?RESPONSE_503)
     end.
 
 -spec prepared_query(Pool::term(), Name::string(), Params::[epgsql:bind_param()]) -> epgsql_cmd_prepared_query:response().
 prepared_query(Pool, Name, Params) ->
-    case do_checkout(Pool, ?POOL_CHECKOUT_RETRIES) of
-        {ok, Reference, Conn} ->
-            Res = epgsql:prepared_query(Conn, Name, Params),
-            dispcount:checkin(Pool, Reference, Conn),
-            Res;
+    Ref = make_ref(),
+    Fun = fun(From, {Stmts, Conn}) ->
+                  Statement = maps:get(Name, Stmts),
+                  #statement{types = Types} = Statement,
+                  TypedParameters = lists:zip(Types, Params),
+                  %% construct the same kind of cast the epgsqla:prepared_statement does, but redirect
+                  %% the output to the elli process directly
+                  gen_server:cast(Conn, {{cast, From, Ref}, epgsql_cmd_prepared_query, {Statement, TypedParameters}})
+          end,
+    case dispcount:transaction(Pool, Fun) of
+        ok ->
+            receive
+                {_Conn, Ref, Res} ->
+                    Res
+            after
+                500 ->
+                    throw(?RESPONSE_503)
+            end;
         {error, busy} ->
             throw(?RESPONSE_503)
     end.
 
 init(Args) ->
-    process_flag(trap_exit, true),
     GetOpt = fun(K) ->
                      case lists:keyfind(K, 1, Args) of
                          false -> error({missing_opt, K});
                          {_, V} -> V
                      end
              end,
-    DBOpts = GetOpt(db_opts),
+    #{host := Host, username := Username, password := Password} = DBOpts = GetOpt(db_opts),
     Codecs = [{epgsql_codec_json, {jiffy, [], [return_maps]}}],
-    {ok, Conn} = epgsql:connect(DBOpts#{codecs => Codecs}),
-    lists:foreach(fun(Mod) ->
-                          Mod:prepare_conn(Conn)
-                  end, GetOpt(db_handlers)),
-    {ok, #state{db_conn=Conn}}.
+    {ok, Conn} = epgsqla:start_link(),
+    %% there's not an async connect that just takes Conn and Opts, for some reason
+    Ref = epgsqla:connect(Conn, Host, Username, Password, DBOpts#{codecs => Codecs}),
+    {ok, #state{db_conn=Conn, conn_ref=Ref, given=true, handlers=GetOpt(db_handlers)}}.
 
 checkout(_From, State = #state{given=true}) ->
     lager:warning("unexpected checkout when already checked out"),
     {error, busy, State};
```

Comment:
Can we remove this now since we’ve never seen it?

Reply:
why? this becomes more likely with the async startup of the connections

Reply:
Does it? I’m not sure how it happens but I believe race conditions are insane

Reply:
yeah, because if we get checked out before we're done initializing we want to fail the request